File: haskell/glue/src/Glue/Lib/Bool/Lt.hs
module Glue.Lib.Bool.Lt where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

lt :: IR Eval
lt = NativeFunc ltImpl

ltImpl :: IR Eval -> Eval (IR Eval)
ltImpl a = pure $ NativeFunc (ltRight a)

ltRight :: IR Eval -> IR Eval -> Eval (IR Eval)
ltRight a b = case (a, b) of
    (Integer na, Integer nb) -> pure . Bool $ na < nb
    (Float na, Float nb) -> pure . Bool $ na < nb
    (Integer na, Float nb) -> pure . Bool $ fromIntegral na < nb
    (Float na, Integer nb) -> pure . Bool $ na < fromIntegral nb
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/bool/lt.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Less than comparison function
/// Mirrors Haskell Glue.Lib.Bool.Lt.lt exactly
final Ir lt = IrNativeFunc(ltImpl);

/// Less than comparison implementation
/// Mirrors Haskell Glue.Lib.Bool.Lt.ltImpl exactly
Eval<Ir> ltImpl(Ir a) {
  return Eval.pure(IrNativeFunc(ltRight(a)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Bool.Lt.ltRight exactly
Eval<Ir> Function(Ir) ltRight(Ir a) {
  return (Ir b) {
    return sequenceAll([eval(a), eval(b)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final na), IrInteger(value: final nb)] => Eval.pure(
          IrBool(na < nb),
        ),
        [IrFloat(value: final na), IrFloat(value: final nb)] => Eval.pure(
          IrBool(na < nb),
        ),
        [IrInteger(value: final na), IrFloat(value: final nb)] => Eval.pure(
          IrBool(na < nb),
        ),
        [IrFloat(value: final na), IrInteger(value: final nb)] => Eval.pure(
          IrBool(na < nb),
        ),
        _ => throwError(wrongArgumentType(['number', 'number'])),
      };
    });
  };
}
